File: src/ssh_manager/services/deployer.py
"""Deploy a key's public half to its target(s) and record it.

Resolution: a key_name maps to the host(s) that reference it (one for
``per_service``, many for ``shared``). Each target's provider does the install
(named adapter → generic ssh → manual), and the result is recorded in the
inventory keyed by fingerprint - the join that later makes rotation a checklist.
"""
from __future__ import annotations

from dataclasses import dataclass, field

from ..core.inventory import Deployment, Inventory, KeyRecord, today
from ..core.manifest import Host, Manifest
from ..platforms.base import Platform
from ..providers.base import Target
from ..providers.registry import resolve as resolve_provider
from ..util import log, net
from ..util.errors import SshManagerError
from ..util.paths import Paths
from .keystore import KeyStore


@dataclass
class DeployRecord:
    target: str
    provider: str
    method: str
    verified: bool
    detail: str = ""
    error: bool = False   # an automated deploy was attempted and failed (or unreachable)


@dataclass
class DeployReport:
    key_name: str
    fingerprint: str
    records: list[DeployRecord] = field(default_factory=list)

    def format(self) -> str:
        lines = [f"deploy {self.key_name}  ({self.fingerprint})"]
        for r in self.records:
            flag = "verified" if r.verified else "MANUAL/needs-redeploy"
            lines.append(f"  -> {r.target} via {r.provider}/{r.method}: {flag}")
            if r.detail:
                lines.append(f"     {r.detail}")
        return "\n".join(lines)


class Deployer:
    def __init__(self, platform: Platform, paths: Paths,
                 manifest: Manifest, inventory: Inventory) -> None:
        self._platform = platform
        self._paths = paths
        self._manifest = manifest
        self._inventory = inventory
        self._keystore = KeyStore(platform)

    def deploy(self, key_name: str, target_alias: str | None = None) -> DeployReport:
        hosts = self._targets(key_name, target_alias)
        profile = self._profile_for(key_name)
        pub = self._paths.ssh_dir / "profiles" / profile / f"{key_name}.pub"
        if not pub.exists():
            raise SshManagerError(f"public key not found: {pub} - run `sshmgr reconcile` first")
        fp = self._keystore.fingerprint(pub)
        rec = self._ensure_record(fp, profile, key_name)

        report = DeployReport(key_name=key_name, fingerprint=fp)
        for host in hosts:
            provider = resolve_provider(host.provider, self._paths.providers)
            # Reachability precheck for SSH-to-host providers: fail fast (with a
            # VPN-aware message) instead of hanging on an unreachable / VPN-gated host.
            if provider.category == "server":
                st = net.check(host.hostname, host.port, ssh=True,
                               requires_vpn=host.requires_vpn, vpn_name=host.vpn_name,
                               vpn_url=host.vpn_url)
                if not st.reachable:
                    self._record_deployment(rec, host.alias, "unreachable", False)
                    report.records.append(DeployRecord(
                        target=host.alias, provider=provider.name, method="unreachable",
                        verified=False, detail=st.message, error=True))
                    log.audit(self._paths.audit_log, "deploy.unreachable",
                              key=key_name, target=host.alias, hostname=host.hostname)
                    continue
            tgt = Target(
                alias=host.alias, hostname=host.hostname, user=host.user, port=host.port,
                pubkey_path=pub, pubkey_text=pub.read_text(encoding="utf-8"),
                token_env=host.token_env,
                known_hosts=self._paths.ssh_dir / "profiles" / profile / "known_hosts",
            )
            outcome = provider.deploy(tgt)
            self._record_deployment(rec, host.alias, outcome.method, outcome.verified)
            report.records.append(DeployRecord(
                target=host.alias, provider=provider.name, method=outcome.method,
                verified=outcome.verified, detail=outcome.detail, error=outcome.error,
            ))
            log.audit(self._paths.audit_log, "deploy", key=key_name, fingerprint=fp,
                      target=host.alias, provider=provider.name,
                      method=outcome.method, verified=outcome.verified)
        return report

    # resolution
    def _targets(self, key_name: str, target_alias: str | None) -> list[Host]:
        using = [
            rk.host for rk in self._manifest.iter_resolved() if rk.key_name == key_name
        ]
        if not using:
            raise SshManagerError(f"no host in the manifest uses key {key_name!r}")
        if target_alias is None:
            return using
        chosen = [h for h in using if h.alias == target_alias]
        if not chosen:
            raise SshManagerError(
                f"host {target_alias!r} does not use key {key_name!r} "
                f"(it's used by: {', '.join(h.alias for h in using)})"
            )
        return chosen

    def _profile_for(self, key_name: str) -> str:
        for rk in self._manifest.iter_resolved():
            if rk.key_name == key_name:
                return rk.profile
        raise SshManagerError(f"no host in the manifest uses key {key_name!r}")

    # inventory
    def _ensure_record(self, fp: str, profile: str, key_name: str) -> KeyRecord:
        rec = self._inventory.keys.get(fp)
        if rec is None:
            rec = KeyRecord(
                profile=profile,
                path=str(self._manifest.identity_file(profile, key_name)),
            )
            self._inventory.record(fp, rec)
        return rec

    def _record_deployment(self, rec: KeyRecord, target: str,
                           method: str, verified: bool) -> None:
        # Idempotent: one entry per target - replace any existing record for it.
        rec.deployments = [d for d in rec.deployments if d.target != target]
        rec.deployments.append(
            Deployment(target=target, method=method, date=today(), verified=verified)
        )

File: internal/services/deployer/deployer.go
// Package deployer installs a key's public half on its target(s) and records the
// deployment, ported from services/deployer.py. A key name maps to the host(s)
// that reference it (one per_service, many shared); each host's provider does the
// install (named adapter -> generic ssh -> manual) and the result is recorded in
// the inventory keyed by fingerprint.
package deployer

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/simtabi/ssh-manager/internal/core/inventory"
	"github.com/simtabi/ssh-manager/internal/core/manifest"
	"github.com/simtabi/ssh-manager/internal/core/providers"
	"github.com/simtabi/ssh-manager/internal/services/keystore"
	"github.com/simtabi/ssh-manager/internal/util/netcheck"
	"github.com/simtabi/ssh-manager/internal/util/paths"
)

// DeployRecord is one target's deploy outcome.
type DeployRecord struct {
	Target   string
	Provider string
	Method   string
	Verified bool
	Detail   string
	Error    bool
}

// DeployReport summarizes a deploy run.
type DeployReport struct {
	KeyName     string
	Fingerprint string
	Records     []DeployRecord
}

// Format renders the deploy summary (mirrors DeployReport.format).
func (r DeployReport) Format() string {
	lines := []string{fmt.Sprintf("deploy %s  (%s)", r.KeyName, r.Fingerprint)}
	for _, rec := range r.Records {
		flag := "MANUAL/needs-redeploy"
		if rec.Verified {
			flag = "verified"
		}
		lines = append(lines, fmt.Sprintf("  -> %s via %s/%s: %s", rec.Target, rec.Provider, rec.Method, flag))
		if rec.Detail != "" {
			lines = append(lines, "     "+rec.Detail)
		}
	}
	return strings.Join(lines, "\n")
}

// AnyError reports whether any target's deploy was attempted and failed.
func (r DeployReport) AnyError() bool {
	for _, rec := range r.Records {
		if rec.Error {
			return true
		}
	}
	return false
}

// Deployer deploys keys to their targets.
type Deployer struct {
	p   paths.Paths
	m   *manifest.Manifest
	inv *inventory.Inventory
	ks  *keystore.KeyStore
}

// New builds a Deployer.
func New(p paths.Paths, m *manifest.Manifest, inv *inventory.Inventory) *Deployer {
	return &Deployer{p: p, m: m, inv: inv, ks: keystore.New()}
}

// Deploy installs key keyName on its target(s) (all hosts using it, or just
// targetAlias) and records each deployment. Mirrors Deployer.deploy.
func (d *Deployer) Deploy(keyName, targetAlias string) (DeployReport, error) {
	hosts, profile, err := d.targets(keyName, targetAlias)
	if err != nil {
		return DeployReport{}, err
	}
	pub := filepath.Join(d.p.SSHDir, "profiles", profile, keyName+".pub")
	if _, err := os.Stat(pub); err != nil {
		return DeployReport{}, fmt.Errorf("public key not found: %s - run `sshmgr reconcile` first", pub)
	}
	fp, err := d.ks.Fingerprint(pub)
	if err != nil {
		return DeployReport{}, err
	}
	pubText := ""
	if b, err := os.ReadFile(pub); err == nil {
		pubText = string(b)
	}
	fpKey := d.ensureRecord(fp, profile, keyName)
	rec := d.inv.Keys[fpKey]

	report := DeployReport{KeyName: keyName, Fingerprint: fp}
	knownHosts := filepath.Join(d.p.SSHDir, "profiles", profile, "known_hosts")
	for _, host := range hosts {
		provider := providers.Resolve(deref(host.Provider), d.p.Providers())
		if provider.Category() == "server" {
			st := netcheck.Check(host.Hostname, host.Port, host.RequiresVPN, deref(host.VPNName), deref(host.VPNURL), 4*time.Second, true)
			if !st.Reachable {
				recordDeployment(&rec, host.Alias, "unreachable", false)
				report.Records = append(report.Records, DeployRecord{
					Target: host.Alias, Provider: provider.Name(), Method: "unreachable",
					Verified: false, Detail: st.Message(), Error: true,
				})
				continue
			}
		}
		tgt := providers.Target{
			Alias: host.Alias, Hostname: host.Hostname, User: host.User, Port: host.Port,
			PubkeyPath: pub, PubkeyText: pubText, TokenEnv: deref(host.TokenEnv),
			KnownHosts: knownHosts,
		}
		outcome := provider.Deploy(tgt)
		recordDeployment(&rec, host.Alias, outcome.Method, outcome.Verified)
		report.Records = append(report.Records, DeployRecord{
			Target: host.Alias, Provider: provider.Name(), Method: outcome.Method,
			Verified: outcome.Verified, Detail: outcome.Detail, Error: outcome.Error,
		})
	}
	d.inv.Keys[fpKey] = rec
	return report, nil
}

func (d *Deployer) targets(keyName, targetAlias string) ([]manifest.Host, string, error) {
	rks, err := d.m.IterResolved()
	if err != nil {
		return nil, "", err
	}
	var using []manifest.Host
	profile := ""
	for _, rk := range rks {
		if rk.KeyName == keyName {
			using = append(using, rk.Host)
			if profile == "" {
				profile = rk.Profile
			}
		}
	}
	if len(using) == 0 {
		return nil, "", fmt.Errorf("no host in the manifest uses key %q", keyName)
	}
	if targetAlias == "" {
		return using, profile, nil
	}
	var chosen []manifest.Host
	var aliases []string
	for _, h := range using {
		aliases = append(aliases, h.Alias)
		if h.Alias == targetAlias {
			chosen = append(chosen, h)
		}
	}
	if len(chosen) == 0 {
		return nil, "", fmt.Errorf("host %q does not use key %q (it's used by: %s)", targetAlias, keyName, strings.Join(aliases, ", "))
	}
	return chosen, profile, nil
}

// ensureRecord makes sure an inventory record exists for fp, with pydantic
// defaults (type ed25519, rotate 365), and returns the fingerprint key.
func (d *Deployer) ensureRecord(fp, profile, keyName string) string {
	if _, ok := d.inv.Keys[fp]; !ok {
		d.inv.Record(fp, inventory.KeyRecord{
			Profile: profile, Path: d.m.IdentityFile(profile, keyName),
			Type: "ed25519", RotateAfterDays: 365,
		})
	}
	return fp
}

// recordDeployment replaces any existing entry for target with a fresh one.
func recordDeployment(rec *inventory.KeyRecord, target, method string, verified bool) {
	kept := rec.Deployments[:0:0]
	for _, dep := range rec.Deployments {
		if dep.Target != target {
			kept = append(kept, dep)
		}
	}
	date := inventory.Today()
	kept = append(kept, inventory.Deployment{Target: target, Method: method, Date: &date, Verified: verified})
	rec.Deployments = kept
}

func deref(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}
